relayer/src/chain/cosmos.rs

```diff
@@ -1231,17 +1231,35 @@ impl ChainEndpoint for CosmosSdkChain {
             .map_err(|_| Error::ics02(ClientError::empty_prefix()))
     }
 
-    /// Query the chain status
-    fn query_status(&self) -> Result<StatusResponse, Error> {
-        crate::time!("query_status");
-        crate::telemetry!(query, self.id(), "query_status");
+    /// Query the application status
+    fn query_application_status(&self) -> Result<StatusResponse, Error> {
+        crate::time!("query_application_status");
+        crate::telemetry!(query, self.id(), "query_application_status");
 
-        let status = self.status()?;
+        // query the chain status
+        let status = self.chain_status()?;
+
+        // query the application status
+        let abci_status = self
+            .block_on(self.rpc_client.abci_info())
+            .map_err(|e| Error::rpc(self.config.rpc_addr.clone(), e))?;
+
+        // has the application been updated with latest block?
+        let time = if abci_status.last_block_height == status.sync_info.latest_block_height {
+            // yes, use the chain latest block time
+            status.sync_info.latest_block_time
+        } else {
+            // no, retrieve the time of the header at application latest height
+            self.block_on(self.rpc_client.commit(abci_status.last_block_height))
+                .map_err(|e| Error::rpc(self.config.rpc_addr.clone(), e))?
+                .signed_header
+                .header
+                .time
+        };
```

Comment:
This is the main change for this PR. It requires 3 RPCs instead of one. And this happens for handling of every IBC send packet event with impact on relayer performance and extra churn on the full node. So I am a bit hesitant to move this forward until performance tests are done and/or get all info in one RPC.
I opened https://github.com/tendermint/tendermint/issues/8248 to get input from tendermint team.

Reply:
I am also wary of increasing the pressure on the node, especially for every `SendPacket` event.

Are we seeing this only when issuing queries manually or is it also happening during relaying? If only the former, I would vote for waiting until we can get away with a single RPC call or until something [like this](https://github.com/tendermint/tendermint/issues/8248#issuecomment-1087480473) is implemented.

Reply:
This is seen during relaying and can cause failures for any query. In one extreme case reported by Mircea (see issue) hermes was not relaying at all on a channel as it was not able to spawn workers due to `connection not found`.
This was fixed with a partial fix, similar to the one in this PR but without the `time` fix (i.e. we would only fix the height to be the application height). When I did some tests with custom chains (with increased block execution times) the lack of `time` fix would sometime cause `header in the future` chain error for client updates.